aten/src/ATen/native/LinearAlgebra.cpp

```diff
@@ -616,6 +616,395 @@ Tensor matrix_rank(const Tensor& self, bool symmetric) {
+
+    // Form an index for matrices of rank r.
+    const auto rank_r_mask = at::where(rank == r);
+    std::vector<TensorIndex> rank_r_indices;
+    // Populate indices of rank r matrices only if
+    // batch dimension is non-empty.
+    if (n_batch_dims) {
+      rank_r_indices = at::native::toVectorOfTensorIndices(rank_r_mask, n_batch_dims + 2);
+    }
+    // It is equivalent to indexing at [rank == r, :]
+    rank_r_indices.push_back(Slice());
+
+    const auto S_rank_r = S.index(rank_r_indices) * r_mask.narrow(-1, 0, S_len);
+    S_rank_restricted.index_put_(rank_r_indices, S_rank_r);
+
+    // Modify rank_r_indices to be able to index matrices.
+    // It is equivalent to indexing at [rank == r, :, :]
+    rank_r_indices.push_back(Slice());
+
+    const auto U_rank_r = U.index(rank_r_indices) * r_mask.narrow(-1, 0, U_dim);
+    U_rank_restricted.index_put_(rank_r_indices, U_rank_r);
+
+    const auto Vh_rank_r = Vh.index(rank_r_indices) * r_mask.narrow(-1, 0, V_dim).unsqueeze(-1);
+    Vh_rank_restricted.index_put_(rank_r_indices, Vh_rank_r);
+  }
+
+  return std::make_tuple(
+      std::move(U_rank_restricted),
+      std::move(S_rank_restricted),
+      std::move(Vh_rank_restricted),
+      std::move(rank)
+  );
+}
+
+std::tuple<Tensor, Tensor, Tensor, Tensor>
+linalg_svd_rank_restricted(
+    const Tensor& input,
+    c10::optional<double> atol,
+    c10::optional<double> rtol,
+    bool full_matrices) {
+  Tensor atol_tensor, rtol_tensor;
+  std::tie(atol_tensor, rtol_tensor) = get_atol_rtol(input, atol, rtol);
+  return at::linalg_svd_rank_restricted(input, atol_tensor, rtol_tensor, full_matrices);
+}
+
+std::tuple<Tensor, Tensor, Tensor, Tensor>
+linalg_svd_rank_restricted(
+    const Tensor& input,
+    double tol,
+    bool full_matrices) {
+  return at::linalg_svd_rank_restricted(input, tol, /*rtol=*/0.0, full_matrices);
+}
+
+std::tuple<Tensor, Tensor, Tensor, Tensor>
+linalg_svd_rank_restricted(
+    const Tensor& input,
+    const Tensor& tol,
+    bool full_matrices) {
+  const auto rtol = at::zeros({}, tol.options());
+  return at::linalg_svd_rank_restricted(input, tol, rtol, full_matrices);
+}
+
+std::tuple<Tensor&, Tensor&, Tensor&, Tensor&>
+linalg_svd_rank_restricted_out(
+    const Tensor& input,
+    const c10::optional<Tensor>& atol_opt,
+    const c10::optional<Tensor>& rtol_opt,
+    bool full_matrices,
+    Tensor& U,
+    Tensor& S,
+    Tensor& Vh,
+    Tensor& rank) {
+  Tensor atol, rtol;
+  std::tie(atol, rtol) = get_atol_rtol(input, atol_opt, rtol_opt, "torch.linalg.svd_rank_restricted");
+
+  checkSameDevice("torch.linalg.svd_rank_revealing", U, input);
+  checkSameDevice("torch.linalg.svd_rank_revealing", S, input);
+  checkSameDevice("torch.linalg.svd_rank_revealing", Vh, input);
+  checkSameDevice("torch.linalg.svd_rank_revealing", rank, input);
+  checkSameDevice("torch.linalg.svd_rank_revealing", atol, input, "atol");
+  checkSameDevice("torch.linalg.svd_rank_revealing", rtol, input, "rtol");
```

Comment:
why was this marked as resolved? @nikitaved

Reply:
I test it against `matrix_rank` that relies on these error messages. See my comment https://github.com/pytorch/pytorch/pull/71126#discussion_r799644593. I think it could be a nice follow-up PR on removing these checks for linalg ops.